examples/connext_dds/network_capture/example3-api-advanced/c/example3_subscriber.c

```diff
@@ -0,0 +1,486 @@
+        DDS_DataReader* reader,
+        const struct DDS_SampleLostStatus *status)
+{
+}
+
+void example3Listener_on_subscription_matched(
+        void* listener_data,
+        DDS_DataReader* reader,
+        const struct DDS_SubscriptionMatchedStatus *status)
+{
+}
+
+void example3Listener_on_data_available(
+        void* listener_data,
+        DDS_DataReader* reader)
+{
+    example3DataReader *example3_reader = NULL;
+    struct example3Seq data_seq = DDS_SEQUENCE_INITIALIZER;
+    struct DDS_SampleInfoSeq info_seq = DDS_SEQUENCE_INITIALIZER;
+    DDS_ReturnCode_t retcode;
+    int i;
+
+    example3_reader = example3DataReader_narrow(reader);
+    if (example3_reader == NULL) {
+        fprintf(stderr, "DataReader narrow error\n");
+        return;
+    }
+
+    retcode = example3DataReader_take(
+            example3_reader,
+            &data_seq,
+            &info_seq,
+            DDS_LENGTH_UNLIMITED,
+            DDS_ANY_SAMPLE_STATE,
+            DDS_ANY_VIEW_STATE, DDS_ANY_INSTANCE_STATE);
+    if (retcode == DDS_RETCODE_NO_DATA) {
+        return;
+    } else if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "take error %d\n", retcode);
+        return;
+    }
+
+    for (i = 0; i < example3Seq_get_length(&data_seq); ++i) {
+        if (DDS_SampleInfoSeq_get_reference(&info_seq, i)->valid_data) {
+            printf("Received data:\n");
+            example3TypeSupport_print_data(
+                    example3Seq_get_reference(&data_seq, i));
+            fflush(stdout);
+        }
+    }
+
+    retcode = example3DataReader_return_loan(
+            example3_reader,
+            &data_seq,
+            &info_seq);
+    if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "return loan error %d\n", retcode);
+    }
+}
+
+static int subscriber_shutdown(
+        DDS_DomainParticipant *participant,
+        DDS_Boolean last_participant)
+{
+    DDS_ReturnCode_t retcode;
+    int status = 0;
+
+    if (participant != NULL) {
+        retcode = DDS_DomainParticipant_delete_contained_entities(participant);
+        if (retcode != DDS_RETCODE_OK) {
+            fprintf(stderr, "delete_contained_entities error %d\n", retcode);
+            status = -1;
+        }
+
+        retcode = DDS_DomainParticipantFactory_delete_participant(
+                DDS_TheParticipantFactory,
+                participant);
+        if (retcode != DDS_RETCODE_OK) {
+            fprintf(stderr, "delete_participant error %d\n", retcode);
+            status = -1;
+        }
+    }
+
+    if (last_participant) {
+        retcode = DDS_DomainParticipantFactory_finalize_instance();
+        if (retcode != DDS_RETCODE_OK) {
+            fprintf(stderr, "finalize_instance error %d\n", retcode);
+            status = -1;
+        }
+    }
+
+    /*
+     * Disable network capture.
+     *
+     * This must be:
+     *   - The last network capture function that is called.
+     */
+    if (!NDDS_Utility_disable_network_capture()) {
```

Comment:
you are calling this function twice for both participants. I guest it is intended to be called just one.

Same feedback in the publisher.c

Reply:
You are right. I moved it inside the `last_participant` condition.